samples/tacos/src/main/kotlin/com/slack/circuit/tacos/OrderTacosCircuit.kt

```diff
@@ -0,0 +1,356 @@
+    val isPreviousVisible: Boolean,
+    val isNextEnabled: Boolean,
+    val isNextVisible: Boolean,
+    val isFooterVisible: Boolean,
+    val eventSink: (Event) -> Unit
+  ) : CircuitUiState
+
+  sealed interface Event : CircuitUiEvent {
+    val indexModifier: Int
+      get() = 0
+
+    object Previous : Event {
+      override val indexModifier: Int = -1
+    }
+    object Next : Event {
+      override val indexModifier: Int = 1
+    }
+    object ProcessOrder : Event {
+      override val indexModifier: Int = 1
+    }
+  }
+}
+
+@Immutable
+data class OrderDetails(
+  val filling: Ingredient = Ingredient(""),
+  val toppings: Set<Ingredient> = persistentSetOf(),
+  val baseCost: BigDecimal = BigDecimal("9.99"), // Default taco price
+  val ingredientsCost: BigDecimal = BigDecimal.ZERO
+)
+
+/** List of wizard steps (in order!) */
+private val orderSteps =
+  persistentListOf(FillingsOrderStep, ToppingsOrderStep, ConfirmationOrderStep, SummaryOrderStep)
+
+internal class OrderTacosPresenter(
+  private val fillingsProducer: OrderStep.StateProducer<FillingsOrderStep.State>,
+  private val toppingsProducer: OrderStep.StateProducer<ToppingsOrderStep.State>,
+  private val confirmationProducer: OrderStep.StateProducer<ConfirmationOrderStep.OrderState>,
+  private val summaryProducer: OrderStep.StateProducer<SummaryOrderStep.SummaryState>,
+) : Presenter<OrderTacosScreen.State> {
+  @Composable override fun present(): OrderTacosScreen.State = presentInternal()
+
+  @Composable
+  internal fun presentInternal(
+    initialStep: OrderStep = FillingsOrderStep,
+    initialOrderDetails: OrderDetails = OrderDetails(),
+  ): OrderTacosScreen.State {
+    var currentStep by remember { mutableStateOf(initialStep) }
+    var orderDetails by remember { mutableStateOf(initialOrderDetails) }
+    var isNextEnabled by remember { mutableStateOf(false) }
+
+    // Generate state for the current order step
+    val stepState =
+      produceState(currentStep, orderDetails) { event ->
+        when (event) {
+          is OrderStep.Validation -> isNextEnabled = event.isOrderValid
+          is OrderStep.UpdateOrder -> orderDetails = updateOrder(event, orderDetails)
+          is OrderStep.Restart -> {
+            currentStep = FillingsOrderStep
+            orderDetails = OrderDetails()
+            isNextEnabled = false
+          }
+        }
+      }
+
+    return OrderTacosScreen.State(
+      headerResId = currentStep.headerResId,
+      orderCost = orderDetails.baseCost + orderDetails.ingredientsCost,
+      stepState = stepState,
+      isPreviousVisible = currentStep.index in 1..2,
+      isNextEnabled = isNextEnabled,
+      isNextVisible = currentStep.index < 2,
+      isFooterVisible = currentStep != SummaryOrderStep
+    ) { navEvent ->
+      // process navigation event from child order step
+      processNavigation(currentStep, navEvent) { currentStep = it }
+    }
+  }
+
+  @Composable
+  private fun produceState(
```

Comment:
I would define each of these variables directly rather than this function. It's also a bit of a confusing name given the existence of `produceState` in compose itself for suspend functions

Reply:
> I would define each of these variables directly rather than this function

I extracted them to a separate function to improve readability. Calling these directly just makes the call site a lot more complicated and messy in my opinion.

> It's also a bit of a confusing name given the existence of produceState in compose itself for suspend functions

Hmm good point. Renamed to `produceOrderStepState`. https://github.com/slackhq/circuit/pull/496/commits/0d0489752bfcb54a7c5591aa271c5cef10531a03